src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelConnection.cs

```diff
@@ -1,25 +1,168 @@
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.Extensions.Logging;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
 {
-    internal class KestrelConnection
+    internal class KestrelConnection : IConnectionHeartbeatFeature, IConnectionCompleteFeature, IConnectionLifetimeNotificationFeature
     {
-        private TaskCompletionSource<object> _executionTcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+        private List<(Action<object> handler, object state)> _heartbeatHandlers;
+        private readonly object _heartbeatLock = new object();
 
-        public KestrelConnection(TransportConnection transportConnection)
+        private Stack<KeyValuePair<Func<object, Task>, object>> _onCompleted;
+        private bool _completed;
+
+        private readonly CancellationTokenSource _connectionClosingCts = new CancellationTokenSource();
+        private readonly TaskCompletionSource<object> _completionTcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        public KestrelConnection(ConnectionContext connectionContext, ILogger logger)
+        {
+            Logger = logger;
+            TransportConnection = connectionContext;
+
+            // Set a connection id if the transport didn't set one
+            TransportConnection.ConnectionId ??= CorrelationIdGenerator.GetNextId();
+            connectionContext.Features.Set<IConnectionHeartbeatFeature>(this);
+            connectionContext.Features.Set<IConnectionCompleteFeature>(this);
+            connectionContext.Features.Set<IConnectionLifetimeNotificationFeature>(this);
+            ConnectionClosedRequested = _connectionClosingCts.Token;
+        }
+
+        private ILogger Logger { get; }
+
+        public ConnectionContext TransportConnection { get; set; }
+        public CancellationToken ConnectionClosedRequested { get; set; }
+        public Task ExecutionTask => _completionTcs.Task;
+
+        public void TickHeartbeat()
+        {
+            lock (_heartbeatLock)
+            {
+                if (_heartbeatHandlers == null)
+                {
+                    return;
+                }
+
+                foreach (var (handler, state) in _heartbeatHandlers)
+                {
+                    handler(state);
+                }
+            }
+        }
+
+        public void OnHeartbeat(Action<object> action, object state)
+        {
+            lock (_heartbeatLock)
+            {
+                if (_heartbeatHandlers == null)
+                {
+                    _heartbeatHandlers = new List<(Action<object> handler, object state)>();
+                }
+
+                _heartbeatHandlers.Add((action, state));
+            }
+        }
+
+        void IConnectionCompleteFeature.OnCompleted(Func<object, Task> callback, object state)
+        {
+            if (_completed)
```

Comment:
Don't we have this same pattern with Request OnCompleted? Can we simplify/consolidate this logic somehow?

Reply:
This isn’t new code. It’s just moved, and it’s the same logic. If you have comments on what you’d like to simplify, it can be applied to both patterns in a later commit